### PR TITLE
Auto-generate documentation for `probnum.config` options

### PR DESCRIPTION
**Pull request by marvinpfoertner**
# In a Nutshell
This PR implements a custom Sphinx extension which generates a table documenting the available configuration options automatically.

# Detailed Description
- [x] Sphinx extension generating configuration option documentation automatically
- [x] add per-option docstrings to `probnum.config` via a required (?) `docstring` argument in the `register` method
- [x] store configuration options as inner dataclasses `Configuration.Option` with fields `key`, `value`, `default_value`, and ~`docstring`~ `description`
- ~[ ] add iterator over registered `Configuration.Option`s~

### CODE REVIEW

File: docs/source/_ext/probnum-config-options.py
Comment (marvinpfoertner) [inline docs/source/_ext/probnum-config-options.py:46]: Here we have to
- [ ] use the default value instead of the value in the documentation
- [ ] read the description from the global `Configuration` object

File: src/probnum/_config.py
Comment (schmidtjonathan) [inline src/probnum/_config.py:46]: We have to use the "default" attribute write access here, since we override `__setattr__`.

Comment (schmidtjonathan) [reply src/probnum/_config.py:46]: And for a reason I haven't quite figured out, yet, using `self.__dict__`, instead of a dictionary that is initialized as an additional class attribute (i.e., here, `self._options_registry`), does not work.
If somebody knows why, I would be very grateful to learn.

**Comment by schmidtjonathan** [comment]
Thanks to @marvinpfoertner for writing a really nice sphinx plugin for automatically generating the docs! I just adapted the rest :)

File: src/probnum/_config.py
Comment (schmidtjonathan) [inline src/probnum/_config.py:52]: Another thing I would like to understand, but I don't:
Using `hasattr(self, key)` here does not work. My only explanation is that it depends on the order in which python registers the methods somehow? If somebody knows, I'd appreciate a brief explanation!

**Comment by marvinpfoertner** [comment]
Okay, so @schmidtjonathan and I broke Python with this thing, but it works so nobody touch it anymore! 🤣

File: docs/source/_ext/probnum-config-options.py
Comment (pnkraemer) [inline docs/source/_ext/probnum-config-options.py:8]: I dont understand what this class does. It does not seem crucial that I do, but would it make sense either way to document the objective and behaviour, at least with a few one-liner docstrings in the methods?

File: docs/source/_ext/probnum-config-options.py
Comment (pnkraemer) [inline docs/source/_ext/probnum-config-options.py:15]: is this on purpose?

File: src/probnum/_config.py
Comment (pnkraemer) [inline src/probnum/_config.py:76]: so all the options are handled inside the `_options_registry`? Is this attribute introduced somewhere? This is no meaningful review comment, I just don't understand where it is defined (and documented?).

**Review by pnkraemer — COMMENTED** [review]
Thanks for doing this! I have two stupid questions (plus a very minor one), so it would be great if you could answer those. Other than that, the PR should be good to go.

File: docs/source/_ext/probnum-config-options.py
Comment (JonathanWenger) [reply docs/source/_ext/probnum-config-options.py:8]: +1 for a short docstring

**Review by JonathanWenger — APPROVED** [review]
Great job for putting in the extra effort here to make this more maintainable. One tiny point: Could you re-order both the sidebar and table under "API Reference" so that `probnum.config` follows the alphabetical order of the subpackages (i.e. is at the top)? If you feel this is out of scope for this PR, that's fine as well.

File: src/probnum/_config.py
Comment (marvinpfoertner) [reply src/probnum/_config.py:76]: `_options_registry` is initialized in the constructor, i.e. via

```
object.__setattr__(self, "_options_registry", dict())
```

This monstrosity is necessary, since we overwrite `__setattr__`, so simply calling

```
self._options_registry = {}
```

sadly doesn't work

File: docs/source/_ext/probnum-config-options.py
Comment (marvinpfoertner) [reply docs/source/_ext/probnum-config-options.py:8]: I can add a docstring tomorrow, but I'm not sure how useful this is, since the docstring might be without reading the Sphinx extension tutorial and having read  the tutorial it might be trivial

File: docs/source/_ext/probnum-config-options.py
Comment (marvinpfoertner) [reply docs/source/_ext/probnum-config-options.py:15]: Yup. In `docutils.nodes`, basically nothing is documented and it took some digging to find out how to adjust column widths automatically. In case we want to switch to statically sized columns, I added the respective command in order to save us the pain in the future

File: docs/source/_ext/probnum-config-options.py
Comment (JonathanWenger) [reply docs/source/_ext/probnum-config-options.py:8]: Even just linking to the tutorial could be useful information for future maintainers.

Comment (pnkraemer) [reply docs/source/_ext/probnum-config-options.py:8]: Thanks! I agree that even only a link to the tutorial might be useful :)

File: docs/source/_ext/probnum-config-options.py
Comment (pnkraemer) [reply docs/source/_ext/probnum-config-options.py:15]: Do you think it makes sense to put a quick comment into the source here that explains exactly this issue? If not, that is okay as well.

File: src/probnum/_config.py
Comment (pnkraemer) [reply src/probnum/_config.py:76]: Okay I didnt see it (maybe because it didnt look like a standard init ;)). Thanks. I will just believe you that it is necessary :D

**Comment by schmidtjonathan** [comment]
@pnkraemer , @JonathanWenger , thanks for your reviews!
I integrated your feedback, do you want to have another look before I merge?

File: docs/source/_ext/probnum-config-options.py
Comment (schmidtjonathan) [reply docs/source/_ext/probnum-config-options.py:26]: In my opinion, the keywords are sufficiently expressive, and a comment is not necessary. But I can add one, if you prefer :)

Comment (pnkraemer) [reply docs/source/_ext/probnum-config-options.py:26]: I think it will do the job without (especially if I would be the only one to whom this ocmment wasnt obvious :))

**Review by pnkraemer — APPROVED** [review]
Thanks guys!

**Comment by JonathanWenger** [comment]
> @pnkraemer , @JonathanWenger , thanks for your reviews!
> 
> I integrated your feedback, do you want to have another look before I merge?

All good! Thanks!

**Pull request merged**
